Scaffold a basic app for node & fetch it via npm.

ngryman opened the pull request:
Related to #4.

alrra commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
 node_modules
-app/templates
+app/templates/apache
```
@ngryman this should be left to `app/templates`

ngryman replied:
I'm using `templates` for copying `package.json` and `index.js` to the project.
Where should I put those files?

ngryman replied:
ping @alrra.

sindresorhus commented on app/index.js:
```diff
@@ -3,6 +3,7 @@
 var fs = require('fs');
 var path = require('path');
 var util = require('util');
+var npm = require("npm");
```
npm is a HUGE dependency. better to use the yeoman API method `installDependencies()`.

arthurvr replied:
Btw, avoid double quotes.

arthurvr commented on app/templates/index.js:
```diff
@@ -0,0 +1,4 @@
+var h5bp = require('h5bp');
+
+var app = h5bp.createServer({ root: __dirname + '/public' });
+app.listen(3000);
```
Missing newline

ngryman replied:
Can you elaborate?

arthurvr replied:
@ngryman Our [`.editorconfig`](https://github.com/h5bp/generator-server-configs/blob/master/.editorconfig#L8) states we want newlines at the end of a file. This file doesn't have one. 
![scaffold_a_basic_app_for_node___fetch_it_via_npm__by_ngryman_ _pull_request__5_ _h5bp_generator-server-configs](https://cloud.githubusercontent.com/assets/6025224/5887323/dd924e16-a3c9-11e4-9dd8-d37281cde679.png)

sindresorhus commented on app/templates/package.json:
```diff
@@ -0,0 +1,16 @@
+	"engines": {
+		"node": ">=0.8.0"
+	}
+}
```
use this package.json instead https://github.com/sindresorhus/node-module-boilerplate/blob/master/boilerplate/package.json

sindresorhus commented on app/index.js:
```diff
@@ -80,13 +80,19 @@ Generator.prototype.copyFiles = function (arg) {
 
-        var configsPath = path.join(this.sourceRoot(), configs.server);
+        // node server config is fetched via npm
+        if ( configs.server == 'node' ) {
```
no space here

sindresorhus commented on app/index.js:
```diff
@@ -159,4 +165,13 @@ Generator.prototype.copyFiles = function (arg) {
 
 };
 
+Generator.prototype.installNodeConfig = function() {
```
`function ()`

sindresorhus commented on app/index.js:
```diff
@@ -159,4 +165,13 @@ Generator.prototype.copyFiles = function (arg) {
 };
 
+Generator.prototype.installNodeConfig = function() {
+  if (!this.isNodeConfig) return;
```
``` js
if (!this.isNodeConfig) {
  return;
}
```

sindresorhus commented on app/index.js:
```diff
@@ -80,13 +80,19 @@ Generator.prototype.copyFiles = function (arg) {
+        // node server config is fetched via npm
+        if ( configs.server == 'node' ) {
+            this.copy('package.json', 'package.json');
+            this.copy('index.js', 'index.js');
```
one argument is enough if the src/dest has the same name

sindresorhus commented:
@ngryman Can you fix the merge conflict and address the inline comments so we can get this merged?

ngryman commented:
@sindresorhus Yes sorry I've been ultra busy, will do :)

arthurvr commented:
ping @ngryman

ngryman commented:
Hi,

After a long long time ;)
- fixes based on your feedback.
- updated to latest generator code.
- used your `package.json` template @sindresorhus. Some tweaks:
  - removed `<% ... %>` from it as it crashes `npm` when installing.
  - removed `files` entry because inadvertent users may forget to add other project's files.
  - removed `mocha` to let them choose what they want.
- increased tests timeout to let `node-sass` compile when installing `server-configs-node`.

But, tests are failing on `iojs` because `node-sass` dependency is too old. I'm about to make a new release of `server-configs-node` with bumped dependencies. So I'll ping you when it's ready, tests will pass. I'll also squash everything when it's ok.

sindresorhus commented on app/templates/package.json:
```diff
@@ -0,0 +1,25 @@
+    "h5bp": "^0.1.2"
+  },
+  "devDependencies": {}
+}
```
Since this is for a server, we can drop everything except for `scripts` and `dependencies`.

ngryman replied:
@sindresorhus Like everything everything? (name, version, ...)

sindresorhus replied:
Yup

arthurvr commented on app/templates/package.json:
```diff
@@ -0,0 +1,8 @@
+  "dependencies": {
+    "h5bp": "^0.1.3"
+  }
+}
```
Let's don't call this file `package.json` as NPM will try to use it and fail. Just call it `_package.json` and copy it to `package.json`.

arthurvr commented on app/templates/index.js:
```diff
@@ -0,0 +1,4 @@
+var h5bp = require('h5bp');
```
Let's `use strict`

arthurvr commented on app/index.js:
```diff
@@ -118,17 +118,26 @@ var Generator = module.exports = yeoman.generators.Base.extend({
+			this.copy('index.js');
+			done();
+
+			this.npmInstall(['h5bp'], {
```
I wonder if we shouldn't use `installDependencies()` here. Also let's support a `--skip-install` option, nearly every yeoman generator has that.

arthurvr commented on app/index.js:
```diff
@@ -118,17 +118,26 @@ var Generator = module.exports = yeoman.generators.Base.extend({
 		var destPath = this.destinationRoot();
 
+		// node server config is fetched via npm
+		if (configs.server == 'node') {
```
triple equal

arthurvr commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 node_modules
 app/templates
```
Just wondering but why do we even have this rule?

arthurvr commented on app/templates/index.js:
```diff
@@ -0,0 +1,4 @@
+var h5bp = require('h5bp');
+
+var app = h5bp.createServer({ root: __dirname + '/public' });
```
There isn't really a thing as a h5bp style guide so I think we should use the code style @sindresorhus prefers. Which is, if I'm correct, `h5bp.createServer({root: __dirname + '/public'});` (the spacing)

arthurvr commented on package.json:
```diff
@@ -18,7 +18,7 @@
   "repository": "h5bp/generator-server-configs",
   "scripts": {
-    "test": "mocha --timeout 50000"
+    "test": "mocha --timeout 60000"
```
Why is this needed? That seems quite a bunch of time.

ngryman replied:
> increased tests timeout to let node-sass compile when installing server-configs-node.

Compilation may take some time, on my machine (1 yo MBP), it took more that 50sec.

arthurvr replied:
Yeah but generally with testing Yeoman generators you don't install the deps, and just use `skip-install` in the tests.

ngryman replied:
Then I don't see the point of this PR ;)
Anyway, we are speaking about +10sec timeout on tests, is this really important?

arthurvr replied:
No, it isn't too important (though it's 10 seconds of productivity gone).

Installing the dependencies should happen when the users uses the generator, though, why is installing the dependencies really needed to run the tests? Generally we don't do that at any yeoman generators.

ngryman replied:
As installing `h5bp` dep is one of the point of this PR, I thought that was a good idea to test if it was successful. If you guys think it's overkill, I can remove it by skipping install in the tests.

arthurvr replied:
cc @sindresorhus for his thoughts. I personally don't think it adds that much value. I do consider tests running for really long a problem though. 10 seconds extra is quite a bunch.

ngryman replied:
PS: testing it revealed that it doesn't work on `iojs`, which is a good thing to know.

arthurvr replied:
@ngryman that's something that should have been tested at the server-configs repo instead :)

ngryman replied:
Yes sure, but it's good to know that a dependency you have doesn't work so you can file a bug and tell your users why it doesn't work ;)

sindresorhus replied:
I agree with @arthurvr. Not the responsibility of this module to test that.

sindresorhus commented:
Tests are failing.

sindresorhus commented:
And why is does `h5bp` have so many dependencies?

`webmake, uglify-js, express, stylus, requirejs, node-sass, less, async, clean-css`

Feels like it's doing way too much.

arthurvr commented:
> Feels like it's doing way too much.

Yup. A magic server doing all the things is pretty much what server-configs-node has always been :( We should refresh it, one day, ..

ngryman committed:
Scaffold a basic app for node & fetch it via npm.
